mlflow/evaluation.py

```diff
@@ -0,0 +1,354 @@
+from typing import Dict, Union
+import entrypoints
+import warnings
+import mlflow
+import hashlib
+import time
+import numpy as np
+import pandas as pd
+import pickle
+from mlflow.exceptions import MlflowException
+from mlflow.utils.file_utils import TempDir
+from mlflow.entities import Metric
+from mlflow.tracking.artifact_utils import get_artifact_uri, _download_artifact_from_uri
+
+
+class EvaluationMetrics(dict):
+    pass
+
+
+class EvaluationArtifacts:
+
+    def __init__(self, location, content=None):
+        self._content = content
+        self._location = location
+
+    def load_content_from_file(self, local_artifact_file):
+        raise NotImplementedError()
+
+    @property
+    def content(self):
+        """
+        The content of the artifact (representation varies)
+        """
+        if self._content is None:
+            with TempDir() as temp_dir:
+                local_artifact_file = temp_dir.path('local_artifact')
+                _download_artifact_from_uri(self._location, local_artifact_file)
+                self._content = self.load_content_from_file(local_artifact_file)
+
+        return self._content
+
+    @property
+    def location(self) -> str:
+        """
+        The location of the artifact
+        """
+        return self._location
+
+    def __getstate__(self, state):
+        state = state.__dict__.copy()
+        # skip pickling artifact content
+        del state['_content']
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+
+
+class EvaluationResult:
+
+    def __init__(self, metrics, artifacts):
+        self._metrics = metrics
+        self._artifacts = artifacts
+
+    @classmethod
+    def load(cls, path):
+        """Load the evaluation results from the specified local filesystem path"""
+        with open(path, 'r') as f:
+            obj = pickle.load(f)
+        return obj
+
+    def save(self, path):
+        """Write the evaluation results to the specified local filesystem path"""
+        # We will likely avoid serializing artifacts themselves, just locations.
+        # Deserialization will resolve locations to artifact contents.
+        with open(path, 'w') as f:
+            pickle.dump(self, f)
```

Comment:
If artifact content has already been loaded, I don't think we want to pickle it. Ideally, it would be nice if we could save the evaluation result as a JSON-formatted file rather than relying on pickle.

Reply:
To assist with the process, let's have each `EvaluationArtifact` implement its own `save()` method that we can call.

Reply:
what do you think of the approach skip pickling _content in here? (so we can still pickle the whole EvaluationResult instance
https://github.com/mlflow/mlflow/pull/5069/files#r751814783

Reply:
I think it would be nice to save all of the content as files. That way, users can easily inspect the files outside of Python, share them across Python versions etc.